src/data/sign.lean

```diff
@@ -0,0 +1,200 @@
+
+instance {α} [has_zero α] [has_one α] [has_neg α] : has_coe_t sign_type α := ⟨cast⟩
+
+/-- `sign_type.cast` as a `mul_with_zero_hom`. -/
+@[simps] def cast_hom {α} [mul_zero_one_class α] [has_distrib_neg α] : sign_type →*₀ α :=
+{ to_fun    := cast,
+  map_zero' := rfl,
+  map_one'  := rfl,
+  map_mul'  := λ x y, by cases x; cases y; simp }
+
+end sign_type
+
+variables {α : Type*}
+
+open sign_type
+
+section preorder
+
+variables [has_zero α] [preorder α] [decidable_rel ((<) : α → α → Prop)] {a : α}
+
+/-- The sign of an element is 1 if it's positive, -1 if negative, 0 otherwise. -/
+def sign : α →o sign_type :=
+⟨λ a, if 0 < a then 1 else if a < 0 then -1 else 0, λ a b h, begin
+  dsimp,
+  split_ifs; try {constructor},
+  { cases lt_irrefl 0 (h_1.trans $ h.trans_lt h_3) },
+  { cases h_2 (h_1.trans_le h) },
+  { cases h_2 (h.trans_lt h_4) }
+  end⟩
+
+lemma sign_apply : sign a = ite (0 < a) 1 (ite (a < 0) (-1) 0) := rfl
+
+@[simp] lemma sign_zero : sign (0 : α) = 0 := by simp [sign_apply]
+@[simp] lemma sign_pos (ha : 0 < a) :  sign a = 1 := by rwa [sign_apply, if_pos]
+@[simp] lemma sign_neg (ha : a < 0) : sign a = -1 := by rwa [sign_apply, if_neg $ asymm ha, if_pos]
+
+end preorder
+
+section linear_order
+
+variables [has_zero α] [linear_order α] {a : α}
+
+lemma sign_ne_zero (h : a ≠ 0) : sign a ≠ 0 :=
+begin
+  contrapose! h,
+  rw sign_apply at h,
+  split_ifs at h,
+  { cases h },
+  { cases h },
+  exact ((lt_trichotomy a 0).resolve_left h_2).resolve_right h_1
+end
+
+end linear_order
+
+section linear_ordered_ring
+
+variables [linear_ordered_ring α] {a b : α}
+
+/- I'm not sure why this is necessary, see https://leanprover.zulipchat.com/#narrow/stream/
+113488-general/topic/type.20class.20inference.20issues/near/276937942 -/
+local attribute [instance] linear_ordered_ring.decidable_lt
+
+-- TODO: determine whether `ring` is required. Current thinking is that it is, for the implication
+-- a < 0 → b < 0 → 0 < a * b.
+/-- `sign` as a `monoid_with_zero_hom` for a nontrivial ordered semiring. Note that linearity
+is required; consider ℂ with the order `z ≤ w` iff they have the same imaginary part and
+`z - w ≤ 0` in the reals; then `1 + i` and `1 - i` are incomparable to zero, and thus we have:
+`0 * 0 = sign (1 + i) * sign (1 - i) ≠ sign 2 = 1`. (`complex.ordered_comm_ring`) -/
+def sign_hom : α →*₀ sign_type :=
+{ to_fun := sign,
+  map_zero' := sign_zero,
+  map_one' := sign_pos zero_lt_one,
+  map_mul' := λ x y, begin
+    rcases lt_trichotomy x 0 with hx | rfl | hx; rcases lt_trichotomy y 0 with hy | rfl | hy,
+    any_goals { simp only [sign_zero, mul_zero, zero_mul] },
+    all_goals { simp only [sign_pos, sign_neg, hx, hy, mul_one, neg_one_mul, neg_neg, one_mul] },
+    { exact sign_pos (mul_pos_of_neg_of_neg hx hy) },
+    { exact sign_neg (mul_neg_of_neg_of_pos hx hy) },
+    { exact sign_neg (mul_neg_of_pos_of_neg hx hy) },
+    { exact sign_pos (mul_pos hx hy) }
```

Comment:
```suggestion
    all_goals { simp only [sign_pos, sign_neg, hx, hy, mul_one, neg_one_mul, neg_neg, one_mul,
      mul_pos_of_neg_of_neg, mul_neg_of_neg_of_pos, mul_neg_of_pos_of_neg, mul_pos] },
```
This should also work.

Reply:
many thanks, I made a slightly golfier version of this!